### PR TITLE
feat(rds): allow turning on IAM authentication for Clusters

### LINKED ISSUE #13722: (rds): Missing EnableIAMDatabaseAuthentication on DatabaseCluster
AWS::RDS::DBCluster has a EnableIAMDatabaseAuthentication property, but the CDK DatabaseCluster does not expose it

### Use Case

Enabling IAM Authentication for Aurora RDS Clusters.

### Proposed Solution

Add it :-) It should exist on DatabaseClusterFromSnapshot, too.

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by BLasan**
Closes #13722

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by BLasan** [comment]
@skinny85 @rix0rrr Hope the build failed not due to the changes

**Comment by BLasan** [comment]
@skinny85 Build succeeded. Please review :)

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 1]
```diff
@@ -88,6 +88,14 @@ interface DatabaseClusterBaseProps {
    */
   readonly deletionProtection?: boolean;
 
+  /**
+   * Whether to enable mapping of AWS Identity and Access Management (IAM) accounts
+   * to database accounts.
+   *
+   * @default false
+   */
+  readonly iamAuthentication?: boolean;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:97]: I'm wondering: why add this property to `DatabaseClusterBaseProps`? This means this is also available for the `DatabaseClusterFromSnapshot` resource. Is this correct? Can `iamAuthentication` be enabled for Clusters created from snapshots?

Maybe this property should be added to `DatabaseClusterProps` instead?

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 2]
```diff
@@ -296,6 +304,7 @@ abstract class DatabaseClusterNew extends DatabaseClusterBase {
   protected readonly newCfnProps: CfnDBClusterProps;
   protected readonly securityGroups: ec2.ISecurityGroup[];
   protected readonly subnetGroup: ISubnetGroup;
+  protected enableIamAuthentication?: boolean;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:307]: What's the point of this field? Can we get rid of it please?

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 3]
```diff
@@ -355,6 +365,7 @@ abstract class DatabaseClusterNew extends DatabaseClusterBase {
       port: props.port ?? clusterEngineBindConfig.port,
       dbClusterParameterGroupName: clusterParameterGroupConfig?.parameterGroupName,
       associatedRoles: clusterAssociatedRoles.length > 0 ? clusterAssociatedRoles : undefined,
+      enableIamDatabaseAuthentication: Lazy.any({ produce: () => this.enableIamAuthentication }),
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:368]: Why use a Lazy here? Why not just `enableIamDatabaseAuthentication: props. enableIamAuthentication`?

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 4]
```diff
@@ -340,6 +349,7 @@ abstract class DatabaseClusterNew extends DatabaseClusterBase {
     const clusterParameterGroup = props.parameterGroup ?? clusterEngineBindConfig.parameterGroup;
     const clusterParameterGroupConfig = clusterParameterGroup?.bindToCluster({});
     this.engine = props.engine;
+    this.enableIamAuthentication = props.iamAuthentication;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:352]: Same question here - why is this new code in `DatabaseClusterNew`? I'd rather have it in `DatabaseCluster`, unless we're 100% certain this works for Clusters created from Snapshots too.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 5]
```diff
@@ -97,6 +97,38 @@ describe('cluster', () => {
...
+      credentials: {
+        username: 'admin',
+        password: cdk.SecretValue.plainText('tooshort'),
+      },
+      instanceProps: {
+        instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE2, ec2.InstanceSize.SMALL),
+        vpc,
+      },
+      iamAuthentication: false,
+    });
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:119]: Let's remove all of the non-required properties from this test which do not serve any purpose. So, this should be:

```suggestion
    new DatabaseCluster(stack, 'Database', {
      engine: DatabaseClusterEngine.AURORA,
      instanceProps: { vpc },
      iamAuthentication: true,
    });
```

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 6]
```diff
@@ -97,6 +97,38 @@ describe('cluster', () => {
...
+
+    // THEN
+    expect(stack).toHaveResource('AWS::RDS::DBCluster', {
+      Engine: 'aurora',
+      DBSubnetGroupName: { Ref: 'DatabaseSubnets56F17B9A' },
+      MasterUsername: 'admin',
+      MasterUserPassword: 'tooshort',
+      EnableIAMDatabaseAuthentication: false,
+      VpcSecurityGroupIds: [{ 'Fn::GetAtt': ['DatabaseSecurityGroup5C91FDCB', 'GroupId'] }],
+    });
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:129]: Let's use `haveResourceLike()`, and omit all of the non-essential properties from this assertion:

```suggestion
    expect(stack).toHaveResource('AWS::RDS::DBCluster', {
      EnableIAMDatabaseAuthentication: true,
    });
```

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 7]
```diff
@@ -97,6 +97,38 @@ describe('cluster', () => {
...
+      instances: 1,
+      credentials: {
+        username: 'admin',
+        password: cdk.SecretValue.plainText('tooshort'),
+      },
+      instanceProps: {
+        instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE2, ec2.InstanceSize.SMALL),
+        vpc,
+      },
+      iamAuthentication: false,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:118]: Let's make this a positive check instead (`iamAuthentication: true`).

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 8]
```diff
@@ -97,6 +97,38 @@ describe('cluster', () => {
     });
   });
 
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:100]: No need for this extra empty line (there should be a single empty line between the tests).

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 9]
```diff
@@ -97,6 +97,38 @@ describe('cluster', () => {
...
+  test('check exposure of EnableIAMDatabaseAuthentication property in DBCluster', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:101]: This description is a little weird. How about

```suggestion
  test('allows setting the `iamAuthentication` property', () => {
```

**Review by skinny85 — CHANGES_REQUESTED** [review]
Thanks for the PR @BLasan ! Some comments/questions below.

**Comment by BLasan** [comment]
@skinny85 Will make the changes ASAP. Thanks for the review :)

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 2 above)
Comment (BLasan) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:307]: Don't we need need to add that field for the db cluster?

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 1 above)
Comment (BLasan) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:97]: Wrote a test case for `DatabaseClusterFromSnapshot`. Worked fine with the existing changes

Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:97]: I understand a unit test work.

Can you try deploying a `DatabaseClusterFromSnapshot` with this property set, and let me know if it deploys without problems?

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:307]: I don't understand the question, sorry.

Where else are you using this field other than this class?

**Comment by BLasan** [comment]
@skinny85 Made the requested changes. Hope is the correct fix for this. Earlier as you mentioned, those configurations need not to be done inside those classes. (Those classes are extended)

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 10]
```diff
@@ -581,6 +591,13 @@ export interface DatabaseClusterFromSnapshotProps extends DatabaseClusterBasePro
    * However, you can use only the ARN to specify a DB instance snapshot.
    */
   readonly snapshotIdentifier: string;
+  /**
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:594]: Missing empty line before this line.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts (hunk 8 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/test/cluster.test.ts:100]: Still not addressed.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 11]
```diff
@@ -1657,6 +1678,28 @@ describe('cluster', () => {
...
+      snapshotIdentifier: 'mySnapshot',
+      iamAuthentication: true,
+    });
+
+    // THEN
+    expect(stack).toHaveResourceLike('AWS::RDS::DBCluster', {
+      EnableIAMDatabaseAuthentication: true,
+    });
+
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:1702]: Please remove these empty lines.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Thanks for your work @BLasan, but I think you misunderstood my comments from earlier.

If it's true that you can set `iamAuthentication` for both `DatabaseCluster` and `DatabaseClusterFromSnapshot`, then it _should_ be done once, in `DatabaseClusterNew` (the common superclass of `DatabaseCluster` and `DatabaseClusterFromSnapshot`).

Can you confirm that you can successfully deploy a `DatabaseClusterFromSnapshot` with `enableIamDatabaseAuthentication` set to `true`, like I asked in https://github.com/aws/aws-cdk/pull/13958#discussion_r610920581, and, if the answer is that it deploys successfully, revert to the first revision of the production code where you did the change in `DatabaseClusterNew`? (Leave the tests as they are right now)

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts (hunk 11 above)
Comment (BLasan) [reply packages/@aws-cdk/aws-rds/test/cluster.test.ts:1702]: sorry for the mistakes made. Will fix them

**Comment by BLasan** [comment]
Not able to test this by creating the stack. Appreciate if someone could test the changes. Thanks !!

**Review by skinny85 — APPROVED** [review]

**Comment by skinny85** [comment]
> Not able to test this by creating the stack. Appreciate if someone could test the changes. Thanks !!

No worries @BLasan. I've tested the PR, and the property works for Clusters created from snapshots too. I've fixed the PR.

**Comment by BLasan** [comment]
> > Not able to test this by creating the stack. Appreciate if someone could test the changes. Thanks !!
> 
> No worries @BLasan. I've tested the PR, and the property works for Clusters created from snapshots too. I've fixed the PR.

Thanks @skinny85  :)

**Review by skinny85 — APPROVED** [review]

**Pull request merged**